core/blockchain/indexer.hpp

```diff
@@ -0,0 +1,246 @@
+      }
+    }
+
+    void remove(const primitives::BlockInfo &block) {
+      map_.erase(block);
+      db_->remove(toKey(block)).value();
+    }
+
+    void filterUnfinalized() {
+      for (auto it = map_.begin(); it != map_.end();) {
+        if (not block_tree_->isFinalized(it->first)) {
+          it = map_.erase(it);
+        } else {
+          ++it;
+        }
+      }
+    }
+
+    primitives::BlockInfo removeUnfinalized() {
+      for (auto it = map_.begin(); it != map_.end();) {
+        if (not block_tree_->isFinalized(it->first)) {
+          it = map_.erase(it);
+        } else {
+          ++it;
+        }
+      }
+      primitives::BlockInfo max{0, block_tree_->getGenesisBlockHash()};
+      auto batch = db_->batch();
+      auto db_cur = db_->cursor();
+      for (db_cur->seekLast().value(); db_cur->isValid();) {
+        auto info = fromKey(*db_cur->key());
+        db_cur->prev().value();
+        if (not block_tree_->isFinalized(info)) {
+          batch->remove(toKey(info)).value();
+        } else if (max < info) {
+          max = info;
+          break;
+        }
+      }
+      batch->commit().value();
+      return max;
+    }
+
+    void writeFinalized(primitives::BlockNumber first,
+                        primitives::BlockNumber last) {
+      auto begin = map_.lower_bound({first, {}}),
+           end = map_.lower_bound({last + 1, {}});
+      auto batch = db_->batch();
+      for (auto it = begin; it != end; ++it) {
+        if (block_tree_->isFinalized(it->first)) {
+          batch->put(toKey(it->first), scale::encode(it->second).value())
+              .value();
+        }
+      }
+      batch->commit().value();
+    }
+
+    using Search = std::pair<primitives::BlockInfo, Indexed<T>>;
+    struct SearchRaw {
+      Search kv;
+      primitives::BlockInfo last;
+    };
+
+    std::optional<SearchRaw> searchRaw(Descent &descent,
+                                       const primitives::BlockInfo &block) {
+      auto map_it = map_.lower_bound(block);
+      while (true) {
+        if (map_it != map_.end() and descent.can(map_it->first)) {
+          if (not map_it->second.inherit) {
+            return SearchRaw{*map_it, map_it->first};
+          }
+          BOOST_ASSERT(map_it->second.prev);
+          auto r = get(*map_it->second.prev);
+          BOOST_ASSERT(r);
+          return SearchRaw{
+              {*map_it->second.prev, std::move(*r)},
+              map_it->first,
+          };
+        }
+        if (map_it == map_.begin()) {
+          break;
+        }
+        --map_it;
+      }
+      auto db_cur = db_->cursor();
+      for (db_cur->seekReverse(toKey(block)).value(); db_cur->isValid();
+           db_cur->prev().value()) {
+        auto info = fromKey(*db_cur->key());
+        if (not descent.can(info)) {
+          continue;
+        }
+        return SearchRaw{
+            {info, scale::decode<Indexed<T>>(*db_cur->value()).value()},
+            info,
+        };
```

Comment:
Isn't it the same as 
```suggestion
        if (descent.can(info)) {
            return SearchRaw{
            {info, scale::decode<Indexed<T>>(*db_cur->value()).value()},
            info,
        };
        }
       
```